### PR TITLE
ASoC: rt5682: Add the soundwire support

### PR DESCRIPTION
**Pull request by oder-chiou**
This patch adds the soundwire support for ALC5682

Signed-off-by: Oder Chiou <oder_chiou@realtek.com>

### CODE REVIEW

File: sound/soc/codecs/rt5682.c [hunk 1]
```diff
@@ -804,10 +819,27 @@ static const struct snd_kcontrol_new rt5682_if1_45_adc_swap_mux =
...
+static SOC_ENUM_SINGLE_DECL(rt5682_dacr_enum,
+	RT5682_AD_DA_MIXER, RT5682_DAC1_R_SEL_SFT, rt5682_dac_select);
+
+static const struct snd_kcontrol_new rt5682_dac_r_mux =
+	SOC_DAPM_ENUM("DAC R Mux", rt5682_dacr_enum);
+
 static void rt5682_reset(struct regmap *regmap)
 {
 	regmap_write(regmap, RT5682_RESET, 0);
+#if !IS_ENABLED(CONFIG_SOUNDWIRE)
```
Comment (bardliao) [inline sound/soc/codecs/rt5682.c]: It may not work if soundwire is enabled by other sdw codec, but rt5682 works on i2s mode. For example, rt5682 works on i2s mode while rt1308 works on sdw mode. Maybe we can use CONFIG_SND_SOC_RT5682_SDW?

File: sound/soc/codecs/rt5682.c [hunk 2]
```diff
@@ -70,9 +79,13 @@ struct rt5682_priv {
...
 
+struct sdw_stream_data {
+	struct sdw_stream_runtime *sdw_stream;
+};
+
 static const struct reg_sequence patch_list[] = {
-	{RT5682_HP_IMP_SENS_CTRL_19, 0x1000},
-	{RT5682_DAC_ADC_DIG_VOL1, 0xa020},
+	{RT5682_HP_IMP_SENS_CTRL_19,	0x1000},
+	{RT5682_DAC_ADC_DIG_VOL1,	0xa020},
```
Comment (bardliao) [inline sound/soc/codecs/rt5682.c]: Should this be separate to different commit?

File: sound/soc/codecs/rt5682.c [hunk 3]
```diff
@@ -2452,7 +2776,42 @@ static const struct snd_soc_component_driver soc_component_dev_rt5682 = {
...
+
+	regmap_write(rt5682->sdw_regmap, RT5682_SDW_CMD, 0);
+	regmap_write(rt5682->sdw_regmap, RT5682_SDW_ADDR_H, (reg >> 8) & 0xff);
+	regmap_write(rt5682->sdw_regmap, RT5682_SDW_ADDR_L, (reg & 0xff));
+	regmap_read(rt5682->sdw_regmap, RT5682_SDW_DATA_H, &data_h);
+	regmap_read(rt5682->sdw_regmap, RT5682_SDW_DATA_L, &data_l);
+
+	*val = (data_h << 8) | data_l;
+
+	dev_dbg(dev, "[%s] %04x => %04x\n", __func__, reg, *val);
```
Comment (bardliao) [inline sound/soc/codecs/rt5682.c]: dev_vdbg

File: sound/soc/codecs/rt5682.c [hunk 4]
```diff
@@ -2452,7 +2776,42 @@ static const struct snd_soc_component_driver soc_component_dev_rt5682 = {
...
+	struct device *dev = context;
+	struct rt5682_priv *rt5682 = dev_get_drvdata(dev);
+
+	regmap_write(rt5682->sdw_regmap, RT5682_SDW_CMD, 1);
+	regmap_write(rt5682->sdw_regmap, RT5682_SDW_ADDR_H, (reg >> 8) & 0xff);
+	regmap_write(rt5682->sdw_regmap, RT5682_SDW_ADDR_L, (reg & 0xff));
+	regmap_write(rt5682->sdw_regmap, RT5682_SDW_DATA_H, (val >> 8) & 0xff);
+	regmap_write(rt5682->sdw_regmap, RT5682_SDW_DATA_L, (val & 0xff));
+
+	dev_dbg(dev, "[%s] %04x <= %04x\n", __func__, reg, val);
```
Comment (bardliao) [inline sound/soc/codecs/rt5682.c]: dev_vdbg

File: sound/soc/codecs/rt5682.c [hunk 5]
```diff
@@ -2725,6 +3091,438 @@ static struct i2c_driver rt5682_i2c_driver = {
...
+
+	regmap_read(rt5682->regmap, RT5682_DEVICE_ID, &val);
+	if (val != DEVICE_ID) {
+		pr_err("Device with ID register %x is not rt5682\n", val);
+		return -ENODEV;
+	}
+
+	rt5682_reset(rt5682->regmap);
+
+	mutex_init(&rt5682->calibrate_mutex);
```
Comment (bardliao) [inline sound/soc/codecs/rt5682.c]: rt5682_io_init() could be called multiple times. So It is better to have a flag to make sure the mutex_init() or INIT_DELAYED_WORK() will be only called once.

File: sound/soc/codecs/rt5682.c [hunk 6]
```diff
@@ -2725,6 +3091,438 @@ static struct i2c_driver rt5682_i2c_driver = {
...
+	/* Regmap Initialization */
+	regmap = devm_regmap_init_sdw(slave, &rt5682_sdw_regmap);
+	if (!regmap)
+		return -EINVAL;
+
+	rt5682_init(&slave->dev, regmap, slave);
+
+	/* Perform IO operations only if slave is in ATTACHED state */
+	if (slave->status == SDW_SLAVE_ATTACHED)
+		rt5682_io_init(&slave->dev, slave);
```
Comment (bardliao) [inline sound/soc/codecs/rt5682.c]: I am not sure if it is required.

File: sound/soc/codecs/rt5682.c [hunk 7]
```diff
@@ -2725,6 +3091,438 @@ static struct i2c_driver rt5682_i2c_driver = {
...
+		dev_err(&slave->dev, "Initialization not complete, timed out\n");
+		return -ETIMEDOUT;
+	}
+
+regmap_sync:
+	slave->unattach_request = 0;
+	regcache_cache_only(rt5682->regmap, false);
+	regcache_sync(rt5682->regmap);
+
+	rt5682_irq(0, rt5682);
```
Comment (bardliao) [inline sound/soc/codecs/rt5682.c]: Do we need to check JD status on resume?

File: sound/soc/codecs/rt5682.c [hunk 8]
```diff
@@ -34,11 +36,13 @@
 
 #define RT5682_NUM_SUPPLIES 3
 
+#if !IS_ENABLED(CONFIG_SOUNDWIRE)
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: ah, this is a problem for Intel.
We have platforms that use rt5682 in I2S mode and we want to use it in SDW mode on others, with the same build/Kconfigs.

File: sound/soc/codecs/rt5682.c [hunk 9]
```diff
@@ -804,10 +819,27 @@ static const struct snd_kcontrol_new rt5682_if1_45_adc_swap_mux =
...
+
+static const struct snd_kcontrol_new rt5682_dac_l_mux =
+	SOC_DAPM_ENUM("DAC L Mux", rt5682_dacl_enum);
+
+static SOC_ENUM_SINGLE_DECL(rt5682_dacr_enum,
+	RT5682_AD_DA_MIXER, RT5682_DAC1_R_SEL_SFT, rt5682_dac_select);
+
+static const struct snd_kcontrol_new rt5682_dac_r_mux =
+	SOC_DAPM_ENUM("DAC R Mux", rt5682_dacr_enum);
+
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: different patch? doesn't looks related to SoundWire, does it?

File: sound/soc/codecs/rt5682.c [hunk 10]
```diff
@@ -879,8 +911,14 @@ static void rt5682_enable_push_button_irq(struct snd_soc_component *component,
...
 			RT5682_4BTN_IL_MASK | RT5682_4BTN_IL_RST_MASK,
 			RT5682_4BTN_IL_EN | RT5682_4BTN_IL_NOR);
+#if !IS_ENABLED(CONFIG_SOUNDWIRE)
 		snd_soc_component_update_bits(component, RT5682_IRQ_CTRL_3,
 			RT5682_IL_IRQ_MASK, RT5682_IL_IRQ_EN);
+#else
+		snd_soc_component_update_bits(component, RT5682_IRQ_CTRL_3,
+			RT5682_IL_IRQ_MASK | RT5682_IL_IRQ_TYPE_MASK,
+			RT5682_IL_IRQ_EN | RT5682_IL_IRQ_PUL);
+#endif
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: I think you need a private field/variable that's set when you probe in SoundWire mode, that would help you select the right path

File: sound/soc/codecs/rt5682.c [hunk 11]
```diff
@@ -972,6 +1010,7 @@ static irqreturn_t rt5682_irq(int irq, void *data)
 	return IRQ_HANDLED;
 }
 
+#if !IS_ENABLED(CONFIG_SOUNDWIRE)
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: remove

File: sound/soc/codecs/rt5682.c [hunk 12]
```diff
@@ -998,6 +1038,7 @@ static int rt5682_set_jack_detect(struct snd_soc_component *component,
 
 	rt5682->hs_jack = hs_jack;
 
+#if !IS_ENABLED(CONFIG_SOUNDWIRE)
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: add private variable set in SoundWire (or I2S mode)

File: sound/soc/codecs/rt5682.c [hunk 13]
```diff
@@ -1133,19 +1175,21 @@ static void rt5682_jack_detect_handler(struct work_struct *work)
 			    SND_JACK_BTN_0 | SND_JACK_BTN_1 |
 			    SND_JACK_BTN_2 | SND_JACK_BTN_3);
 
+#if !IS_ENABLED(CONFIG_SOUNDWIRE)
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: private variable

File: sound/soc/codecs/rt5682.c [hunk 14]
```diff
@@ -1133,19 +1175,21 @@ static void rt5682_jack_detect_handler(struct work_struct *work)
...
+#endif
 
 	mutex_unlock(&rt5682->calibrate_mutex);
 }
 
 static const struct snd_kcontrol_new rt5682_snd_controls[] = {
 	/* DAC Digital Volume */
 	SOC_DOUBLE_TLV("DAC1 Playback Volume", RT5682_DAC1_DIG_VOL,
-		RT5682_L_VOL_SFT + 1, RT5682_R_VOL_SFT + 1, 86, 0, dac_vol_tlv),
+		RT5682_L_VOL_SFT + 1, RT5682_R_VOL_SFT + 1, 87, 0, dac_vol_tlv),
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: different patch?

File: sound/soc/codecs/rt5682.c [hunk 15]
```diff
@@ -1221,6 +1265,7 @@ static int set_dmic_clk(struct snd_soc_dapm_widget *w,
 	return 0;
 }
 
+#if !IS_ENABLED(CONFIG_SOUNDWIRE)
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: remove

File: sound/soc/codecs/rt5682.c [hunk 16]
```diff
@@ -1684,9 +1730,14 @@ static const struct snd_soc_dapm_widget rt5682_dapm_widgets[] = {
 		&rt5682_if1_adc_slot_mux),
 
 	/* ADC Mixer */
+#if IS_ENABLED(CONFIG_SOUNDWIRE)
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: private variable

File: sound/soc/codecs/rt5682.c [hunk 17]
```diff
@@ -1707,6 +1758,8 @@ static const struct snd_soc_dapm_widget rt5682_dapm_widgets[] = {
 	SND_SOC_DAPM_PGA("IF1 DAC1", SND_SOC_NOPM, 0, 0, NULL, 0),
 	SND_SOC_DAPM_PGA("IF1 DAC1 L", SND_SOC_NOPM, 0, 0, NULL, 0),
 	SND_SOC_DAPM_PGA("IF1 DAC1 R", SND_SOC_NOPM, 0, 0, NULL, 0),
+	SND_SOC_DAPM_PGA("SOUND DAC L", SND_SOC_NOPM, 0, 0, NULL, 0),
+	SND_SOC_DAPM_PGA("SOUND DAC R", SND_SOC_NOPM, 0, 0, NULL, 0),
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: different patch?

File: sound/soc/codecs/rt5682.c [hunk 18]
```diff
@@ -1723,12 +1776,19 @@ static const struct snd_soc_dapm_widget rt5682_dapm_widgets[] = {
 	SND_SOC_DAPM_MUX("ADCDAT Mux", SND_SOC_NOPM, 0, 0,
 			&rt5682_adcdat_pin_ctrl),
 
+	SND_SOC_DAPM_MUX("DAC L Mux", SND_SOC_NOPM, 0, 0,
+			&rt5682_dac_l_mux),
+	SND_SOC_DAPM_MUX("DAC R Mux", SND_SOC_NOPM, 0, 0,
+			&rt5682_dac_r_mux),
+
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: different patch?

File: sound/soc/codecs/rt5682.c [hunk 19]
```diff
@@ -1744,9 +1804,14 @@ static const struct snd_soc_dapm_widget rt5682_dapm_widgets[] = {
 		&rt5682_alg_dac_r1_mux),
 
 	/* DAC Mixer */
+#if IS_ENABLED(CONFIG_SOUNDWIRE)
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: private variable

File: sound/soc/codecs/rt5682.c [hunk 20]
```diff
@@ -1882,8 +1947,8 @@ static const struct snd_soc_dapm_route rt5682_dapm_routes[] = {
 	{"IF1_ADC Mux", "Slot 2", "IF1 23 ADC Swap Mux"},
 	{"IF1_ADC Mux", "Slot 4", "IF1 45 ADC Swap Mux"},
 	{"IF1_ADC Mux", "Slot 6", "IF1 67 ADC Swap Mux"},
-	{"IF1_ADC Mux", NULL, "I2S1"},
 	{"ADCDAT Mux", "ADCDAT1", "IF1_ADC Mux"},
+	{"AIF1TX", NULL, "I2S1"},
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: different patch?

File: sound/soc/codecs/rt5682.c [hunk 21]
```diff
@@ -1892,17 +1957,35 @@ static const struct snd_soc_dapm_route rt5682_dapm_routes[] = {
...
+	{"SOUND DAC L", NULL, "PLL2F"},
+	{"SOUND DAC R", NULL, "SDWRX"},
+	{"SOUND DAC R", NULL, "DAC Stereo1 Filter"},
+	{"SOUND DAC R", NULL, "PLL2B"},
+	{"SOUND DAC R", NULL, "PLL2F"},
+
+	{"DAC L Mux", "IF1", "IF1 DAC1 L"},
+	{"DAC L Mux", "SOUND", "SOUND DAC L"},
+	{"DAC R Mux", "IF1", "IF1 DAC1 R"},
+	{"DAC R Mux", "SOUND", "SOUND DAC R"},
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: different patch?

File: sound/soc/codecs/rt5682.c [hunk 22]
```diff
@@ -1892,17 +1957,35 @@ static const struct snd_soc_dapm_route rt5682_dapm_routes[] = {
...
+
 	{"DAC1 MIXL", "Stereo ADC Switch", "Stereo1 ADC MIXL"},
-	{"DAC1 MIXL", "DAC1 Switch", "IF1 DAC1 L"},
+	{"DAC1 MIXL", "DAC1 Switch", "DAC L Mux"},
 	{"DAC1 MIXR", "Stereo ADC Switch", "Stereo1 ADC MIXR"},
-	{"DAC1 MIXR", "DAC1 Switch", "IF1 DAC1 R"},
+	{"DAC1 MIXR", "DAC1 Switch", "DAC R Mux"},
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: different patch for the two above?

File: sound/soc/codecs/rt5682.c [hunk 23]
```diff
@@ -2399,6 +2482,226 @@ static const struct snd_soc_dai_ops rt5682_aif2_dai_ops = {
 	.set_bclk_ratio = rt5682_set_bclk_ratio,
 };
 
+#if IS_ENABLED(CONFIG_SOUNDWIRE)
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: CONFIG_RT5682_SDW?

File: sound/soc/codecs/rt5682.c [hunk 24]
```diff
@@ -2431,6 +2734,27 @@ static struct snd_soc_dai_driver rt5682_dai[] = {
 		},
 		.ops = &rt5682_aif2_dai_ops,
 	},
+#if IS_ENABLED(CONFIG_SOUNDWIRE)
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: CONFIG_RT5682_SDW

File: sound/soc/codecs/rt5682.c [hunk 25]
```diff
@@ -2452,7 +2776,42 @@ static const struct snd_soc_component_driver soc_component_dev_rt5682 = {
 	.endianness		= 1,
 	.non_legacy_dai_naming	= 1,
 };
+#if IS_ENABLED(CONFIG_SOUNDWIRE)
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: CONFIG_RT5682_SDW

File: sound/soc/codecs/rt5682.c [hunk 26]
```diff
@@ -2464,6 +2823,10 @@ static const struct regmap_config rt5682_regmap = {
 	.num_reg_defaults = ARRAY_SIZE(rt5682_reg),
 	.use_single_read = true,
 	.use_single_write = true,
+#if IS_ENABLED(CONFIG_SOUNDWIRE)
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: CONFIG_RT5682_SDW

File: sound/soc/codecs/rt5682.c [hunk 27]
```diff
@@ -2541,6 +2906,7 @@ static void rt5682_calibrate(struct rt5682_priv *rt5682)
 
 }
 
+#if !IS_ENABLED(CONFIG_SOUNDWIRE)
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.c]: remove, keep both I2C available in all cases

File: sound/soc/codecs/rt5682.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/codecs/rt5682.c]: remove as for all other drivers

File: sound/soc/codecs/rt5682.h [hunk 28]
```diff
@@ -177,7 +177,7 @@
 #define RT5682_TEST_MODE_CTRL_4			0x0148
 #define RT5682_TEST_MODE_CTRL_5			0x0149
 #define RT5682_PLL1_INTERNAL			0x0150
-#define RT5682_PLL2_INTERNAL			0x0151
+#define RT5682_PLL2_INTERNAL			0x0156
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.h]: different patch?

File: sound/soc/codecs/rt5682.h [hunk 29]
```diff
@@ -1091,11 +1096,17 @@
 #define RT5682_JD1_POL_MASK			(0x1 << 13)
 #define RT5682_JD1_POL_NOR			(0x0 << 13)
 #define RT5682_JD1_POL_INV			(0x1 << 13)
+#define RT5682_JD1_IRQ_MASK			(0x1 << 10)
+#define RT5682_JD1_IRQ_LEV			(0x0 << 10)
+#define RT5682_JD1_IRQ_PUL			(0x1 << 10)
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.h]: different patch?

File: sound/soc/codecs/rt5682.h [hunk 30]
```diff
@@ -1091,11 +1096,17 @@
...
 
 /* IRQ Control 3 (0x00b8) */
 #define RT5682_IL_IRQ_MASK			(0x1 << 7)
 #define RT5682_IL_IRQ_DIS			(0x0 << 7)
 #define RT5682_IL_IRQ_EN			(0x1 << 7)
+#define RT5682_IL_IRQ_TYPE_MASK			(0x1 << 4)
+#define RT5682_IL_IRQ_LEV			(0x0 << 4)
+#define RT5682_IL_IRQ_PUL			(0x1 << 4)
```
Comment (plbossart) [inline sound/soc/codecs/rt5682.h]: different patch?

File: sound/soc/codecs/rt5682.c (hunk 9 above)
Comment (oder-chiou) [reply sound/soc/codecs/rt5682.c]: It is related with SoundWire, because the Mux is selected the source from the I2S or SoundWire.

File: sound/soc/codecs/rt5682.c (hunk 17 above)
Comment (oder-chiou) [reply sound/soc/codecs/rt5682.c]: This is the source interface from the SoundWire.

File: sound/soc/codecs/rt5682.c (hunk 18 above)
Comment (oder-chiou) [reply sound/soc/codecs/rt5682.c]: This is the source interface from the SoundWire.

File: sound/soc/codecs/rt5682.c (hunk 21 above)
Comment (oder-chiou) [reply sound/soc/codecs/rt5682.c]: This is the source interface from the SoundWire.

File: sound/soc/codecs/rt5682.c (hunk 22 above)
Comment (oder-chiou) [reply sound/soc/codecs/rt5682.c]: This is the source interface from the SoundWire.

File: sound/soc/codecs/rt5682.c (hunk 24 above)
Comment (oder-chiou) [reply sound/soc/codecs/rt5682.c]: Yes, we will modify as that.

File: sound/soc/codecs/rt5682.h (hunk 28 above)
Comment (oder-chiou) [reply sound/soc/codecs/rt5682.h]: The PLL2 is used for SoundWire interface, and the PLL2 register address was wrong.

File: sound/soc/codecs/rt5682.h (hunk 29 above)
Comment (oder-chiou) [reply sound/soc/codecs/rt5682.h]: The settings are for the interrupt of SoundWire interface.

File: sound/soc/codecs/rt5682.h (hunk 30 above)
Comment (oder-chiou) [reply sound/soc/codecs/rt5682.h]: The settings are for the interrupt of SoundWire interface.

File: sound/soc/codecs/rt5682.c (hunk 20 above)
Comment (oder-chiou) [reply sound/soc/codecs/rt5682.c]: The original routing path will enable the widget "I2S1" during SoundWire widget "SDWTX" is enabled. So we moved the widget "I2S1" to connect with the I2S interface "AIF1TX".
